Switching focus from a UITextField to an editable WKWebView causes the keyboard to dance

whsieh opened the pull request:
#### 5df7702c29e60e9fa8676902fc9658747b0f79a9
<pre>
Switching focus from a UITextField to an editable WKWebView causes the keyboard to dance
<a href="https://bugs.webkit.org/show_bug.cgi?id=254236">https://bugs.webkit.org/show_bug.cgi?id=254236</a>
rdar://103712135

Reviewed by Aditya Keerthi.

Currently, when focusing editable web content while a native text field is focused, the keyboard
begins animated dismissal for a few frames, before appearing again. This happens because the web
view steals first responder away from the native text field upon tapping, underneath
`-_singleTapRecognized:`; however, we don&apos;t attempt to reload input views (and thus show the
keyboard) until the element focus IPC message arrives in the UI process (which is sent when
dispatching a synthetic click event). This is partly mitigated by 208968@main and its predecessor,
186343@main, which adopt UIKit SPI in order to keep the keyboard stable in this scenario. While this
works well in practice, it&apos;s still racy in some cases. Consider the following sequence of events:

(UI)    `-_singleTapRecognized:` is called, we become first responder, and begin holding onto the
        input view update deferral token. An activity state update is sent to the web process.

(WEB)   The potential tap is received, and we dispatch onto the runloop to complete the synthetic
        click (and dispatch click events).

(WEB)   The activity state is received, the web view becomes focused, and we tell the UI process
        that the activity state was received.

(WEB)   The synthetic click is dispatched, the editable element is focused, and we send
        `WebPageProxy::ElementDidFocus` to the UI process.

(UI)    The activity state change handler is invoked, causing us to relinquish the input view update
        deferral token. The keyboard begins animating away, since the web view isn&apos;t considered
        editable yet.

(UI)    `WebPageProxy::elementDidFocus()` is called, and we reload input views, causing the keyboard
        to cancel its hiding animation and reappear.

The longer the delay between the last two steps, the more noticeable the &quot;keyboard dance&quot; becomes.
To fix this, we refactor this code so that instead of taking a RAII token representing the scope of
input view update deferral, we instead maintain an `OptionSet` of flags (detailed below) indicating
different reasons for deferring input view updates; as long as any bit in this option set is set, we
defer input view updates.

Test: fast/forms/ios/keyboard-stability-when-tapping-editable-element.html

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/InputViewUpdateDeferrer.h: Removed.
* Source/WebKit/UIProcess/ios/InputViewUpdateDeferrer.mm: Removed.

Delete `InputViewUpdateDeferrer`, now that logic is all in `WKContentViewInteraction`.

* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didHandleTapAsHover):

Add a new IPC message so that the web process can inform the UI process when a tap is handled as a
mouse hover, due to content change observation. See below for more details.

* Source/WebKit/UIProcess/ios/WKContentView.mm:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView cleanUpInteraction]):
(-[WKContentView _cancelPreviousResetInputViewDeferralRequest]):
(-[WKContentView _scheduleResetInputViewDeferralAfterBecomingFirstResponder]):
(-[WKContentView becomeFirstResponderForWebView]):
(-[WKContentView endEditingAndUpdateFocusAppearanceWithReason:]):
(-[WKContentView _commitPotentialTapFailed]):
(-[WKContentView _didNotHandleTapAsClick:]):
(-[WKContentView _didHandleTapAsHover]):
(-[WKContentView _didCompleteSyntheticClick]):
(-[WKContentView _singleTapRecognized:]):
(-[WKContentView startDeferringInputViewUpdates:]):
(-[WKContentView stopDeferringInputViewUpdates:]):
(-[WKContentView stopDeferringInputViewUpdatesForAllSources]):
(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):
(-[WKContentView _resetInputViewDeferral]): Deleted.

Adjust code that previously relied on `InputViewUpdateDeferrer` to instead add or remove
`InputViewUpdateDeferralSources`. There are three distinct sources:

•   `BecomeFirstResponder`: this is first set when becoming first responder in `WKContentView`, and
    is unset once the activity state change (as a result of becoming first responder) has been
    handled by the web process. It&apos;s also unset after the next `ElementDidFocus` message arrives in
    the UI process.

•   `TapGesture`: this is set when handling a tap over web content, right before committing the
    potential tap to the web page; it&apos;s unset once any of the following are invoked:
    `_commitPotentialTapFailed`, `_didNotHandleTapAsClick:`, or `_didCompleteSyntheticClick`.

•   `ChangingFocusedElement`: this is set only in the case where we&apos;re immediately changing focus
    from one element to another underneath `-_elementDidFocus:…:`, and is immediately unset once we
    exit that method.

* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::didHandleTapAsHover):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::handleSyntheticClick):
(WebKit::WebPage::didFinishContentChangeObserving):

Update these to send `WebPageProxy::DidHandleTapAsHover` in the case where the content change
observer causes us to handle the tap as mouse hover, to ensure that we don&apos;t leave the input view
update deferral flag  Also, drive-by fix a typo that I stumbled into
while debugging this code.

* LayoutTests/fast/forms/ios/keyboard-stability-when-tapping-editable-element-expected.txt: Added.
* LayoutTests/fast/forms/ios/keyboard-stability-when-tapping-editable-element.html: Added.
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.keyboardWillHideCount):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView startDeferringInputViewUpdates:]):
(-[WKContentView stopDeferringInputViewUpdates:]):
(-[WKContentView stopDeferringInputViewUpdatesForAllSources]):
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::keyboardWillHideCount const):
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView initWithFrame:configuration:]):
(-[TestRunnerWKWebView _keyboardWillHide]):
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::keyboardWillHideCount const):

Add a new layout test to exercise the change, along with a new testing hook that allows us to check
whether or not there was an attempt to hide the keyboard between any two given points in time (by
simply keeping a tally of the number of times the `KeyboardWillHide` notification was received).

Canonical link: <a href="https://commits.webkit.org/261964@main">https://commits.webkit.org/261964@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b0bb783c2951f54d1762150455d5971e2654c7b5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/83 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/84 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/64 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/77 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/76 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/60 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/70 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/76 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/67 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/61 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/73 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71 "11 api tests failed or timed out") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/70 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/60 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/56 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/74 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/fast/forms/ios/keyboard-stability-when-tapping-editable-element.html:
```diff
@@ -0,0 +1,45 @@
+
+    // Then tap to focus the editable body.
+    await UIHelper.activateElement(document.body);
+    await UIHelper.delayFor(100);
```
Can we write the test without a `UIHelper.delayFor`?

What if we were to check `UIHelper.keyboardWillHideCount` after `UIHelper.waitForKeyboardToHide` and assert that the count only went up by one?

whsieh replied:
Ah, that's a great idea — changed to the following:

```
…
    // Then tap to focus the editable body.
    await UIHelper.activateElement(document.body);

    document.body.blur();
    await UIHelper.waitForKeyboardToHide();

    keyboardWillHideCountAfter = await UIHelper.keyboardWillHideCount();
    shouldBe("keyboardWillHideCountAfter - keyboardWillHideCountBefore", "1");

    finishJSTest();
…
```

Thanks for the review!

whsieh committed:
Switching focus from a UITextField to an editable WKWebView causes the keyboard to dance
https://bugs.webkit.org/show_bug.cgi?id=254236
rdar://103712135

Reviewed by Aditya Keerthi.

Currently, when focusing editable web content while a native text field is focused, the keyboard
begins animated dismissal for a few frames, before appearing again. This happens because the web
view steals first responder away from the native text field upon tapping, underneath
`-_singleTapRecognized:`; however, we don't attempt to reload input views (and thus show the
keyboard) until the element focus IPC message arrives in the UI process (which is sent when
dispatching a synthetic click event). This is partly mitigated by 208968@main and its predecessor,
186343@main, which adopt UIKit SPI in order to keep the keyboard stable in this scenario. While this
works well in practice, it's still racy in some cases. Consider the following sequence of events:

(UI)    `-_singleTapRecognized:` is called, we become first responder, and begin holding onto the
        input view update deferral token. An activity state update is sent to the web process.

(WEB)   The potential tap is received, and we dispatch onto the runloop to complete the synthetic
        click (and dispatch click events).

(WEB)   The activity state is received, the web view becomes focused, and we tell the UI process
        that the activity state was received.

(WEB)   The synthetic click is dispatched, the editable element is focused, and we send
        `WebPageProxy::ElementDidFocus` to the UI process.

(UI)    The activity state change handler is invoked, causing us to relinquish the input view update
        deferral token. The keyboard begins animating away, since the web view isn't considered
        editable yet.

(UI)    `WebPageProxy::elementDidFocus()` is called, and we reload input views, causing the keyboard
        to cancel its hiding animation and reappear.

The longer the delay between the last two steps, the more noticeable the "keyboard dance" becomes.
To fix this, we refactor this code so that instead of taking a RAII token representing the scope of
input view update deferral, we instead maintain an `OptionSet` of flags (detailed below) indicating
different reasons for deferring input view updates; as long as any bit in this option set is set, we
defer input view updates.

Test: fast/forms/ios/keyboard-stability-when-tapping-editable-element.html

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/InputViewUpdateDeferrer.h: Removed.
* Source/WebKit/UIProcess/ios/InputViewUpdateDeferrer.mm: Removed.

Delete `InputViewUpdateDeferrer`, now that logic is all in `WKContentViewInteraction`.

* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didHandleTapAsHover):

Add a new IPC message so that the web process can inform the UI process when a tap is handled as a
mouse hover, due to content change observation. See below for more details.

* Source/WebKit/UIProcess/ios/WKContentView.mm:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView cleanUpInteraction]):
(-[WKContentView _cancelPreviousResetInputViewDeferralRequest]):
(-[WKContentView _scheduleResetInputViewDeferralAfterBecomingFirstResponder]):
(-[WKContentView becomeFirstResponderForWebView]):
(-[WKContentView endEditingAndUpdateFocusAppearanceWithReason:]):
(-[WKContentView _commitPotentialTapFailed]):
(-[WKContentView _didNotHandleTapAsClick:]):
(-[WKContentView _didHandleTapAsHover]):
(-[WKContentView _didCompleteSyntheticClick]):
(-[WKContentView _singleTapRecognized:]):
(-[WKContentView startDeferringInputViewUpdates:]):
(-[WKContentView stopDeferringInputViewUpdates:]):
(-[WKContentView stopDeferringInputViewUpdatesForAllSources]):
(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):
(-[WKContentView _resetInputViewDeferral]): Deleted.

Adjust code that previously relied on `InputViewUpdateDeferrer` to instead add or remove
`InputViewUpdateDeferralSources`. There are three distinct sources:

•   `BecomeFirstResponder`: this is first set when becoming first responder in `WKContentView`, and
    is unset once the activity state change (as a result of becoming first responder) has been
    handled by the web process. It's also unset after the next `ElementDidFocus` message arrives in
    the UI process.

•   `TapGesture`: this is set when handling a tap over web content, right before committing the
    potential tap to the web page; it's unset once any of the following are invoked:
    `_commitPotentialTapFailed`, `_didNotHandleTapAsClick:`, or `_didCompleteSyntheticClick`.

•   `ChangingFocusedElement`: this is set only in the case where we're immediately changing focus
    from one element to another underneath `-_elementDidFocus:…:`, and is immediately unset once we
    exit that method.

* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::didHandleTapAsHover):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::handleSyntheticClick):
(WebKit::WebPage::didFinishContentChangeObserving):

Update these to send `WebPageProxy::DidHandleTapAsHover` in the case where the content change
observer causes us to handle the tap as mouse hover, to ensure that we don't leave the input view
update deferral flag  Also, drive-by fix a typo that I stumbled into
while debugging this code.

* LayoutTests/fast/forms/ios/keyboard-stability-when-tapping-editable-element-expected.txt: Added.
* LayoutTests/fast/forms/ios/keyboard-stability-when-tapping-editable-element.html: Added.
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.keyboardWillHideCount):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView startDeferringInputViewUpdates:]):
(-[WKContentView stopDeferringInputViewUpdates:]):
(-[WKContentView stopDeferringInputViewUpdatesForAllSources]):
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::keyboardWillHideCount const):
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView initWithFrame:configuration:]):
(-[TestRunnerWKWebView _keyboardWillHide]):
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::keyboardWillHideCount const):

Add a new layout test to exercise the change, along with a new testing hook that allows us to check
whether or not there was an attempt to hide the keyboard between any two given points in time (by
simply keeping a tally of the number of times the `KeyboardWillHide` notification was received).

Canonical link: https://commits.webkit.org/261964@main
